README.md
# rest-more [![Build Status](https://secure.travis-ci.org/godfat/rest-more.png?branch=master)](http://travis-ci.org/godfat/rest-more) [![Coverage Status](https://coveralls.io/repos/godfat/rest-more/badge.png)](https://coveralls.io/r/godfat/rest-more) [![Join the chat at https://gitter.im/godfat/rest-more](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/godfat/rest-more)

by Lin Jen-Shin ([godfat](http://godfat.org))

Lin Jen-Shin ([godfat][]) had given a talk about rest-core on
[RubyConf Taiwan 2011][talk]. The slide is in English, but the
talk is in Mandarin. There's another talk about [The Promise of rest-core][]

[godfat]: https://github.com/godfat
[talk]: http://rubyconf.tw/2011/#6
[The Promise of rest-core]: http://godfat.org/slide/2015-01-13-rest-core-promise/

## LINKS:

* [github](https://github.com/godfat/rest-more)
* [rubygems](https://rubygems.org/gems/rest-more)
* [rdoc](http://rdoc.info/projects/godfat/rest-more)
* [mailing list](http://www.freelists.org/list/rest-core)
  Send your questions to: <rest-core@freelists.org> and you could read
  through [archives](http://www.freelists.org/archives/rest-core)

## DESCRIPTION:

Various REST clients such as Facebook and Twitter built with [rest-core][].

[rest-core]: https://github.com/godfat/rest-core

## FEATURES:

Out-of-box REST clients built with rest-core for:

* [RC::Dropbox][] (via OAuth 1.0a)
* [RC::Facebook][] (via OAuth 2, most completed)
* [RC::Github][] (via OAuth 2)
* [RC::Instagram][] (via OAuth 2)
* [RC::Linkedin][] (via OAuth 1.0a)
* [RC::StackExchange][] (via OAuth 2)
* [RC::Twitter][] (via OAuth 1.0a)

[RC::Dropbox]: lib/rest-core/client/dropbox.rb
[RC::Facebook]: lib/rest-core/client/facebook.rb
[RC::Github]: lib/rest-core/client/github.rb
[RC::Instagram]: lib/rest-core/client/instagram.rb
[RC::Linkedin]: lib/rest-core/client/linkedin.rb
[RC::StackExchange]: lib/rest-core/client/stackexchange.rb
[RC::Twitter]: lib/rest-core/client/twitter.rb

Rails utilities are also included for some clients.

Other clients in other gems:

* Firebase: [rest-firebase][]
* TopCoder: [topcoder][]
* YahooBuy: [rest-more-yahoo_buy][]
* Scrapyd: [rest-scrapyd][]

[rest-firebase]: https://github.com/CodementorIO/rest-firebase
[topcoder]: https://github.com/miaout17/topcoder
[rest-more-yahoo_buy]: https://github.com/GoodLife/rest-more-yahoo_buy
[rest-scrapyd]: https://github.com/wvengen/rest-scrapyd

## REQUIREMENTS:

### Mandatory:

* Tested with MRI (official CRuby), Rubinius and JRuby.
* gem [rest-core][]
* gem [httpclient][]
* gem [mime-types][]
* gem [timers][]

[httpclient]: https://github.com/nahi/httpclient
[mime-types]: https://github.com/halostatue/mime-types
[timers]: https://github.com/celluloid/timers

### Optional:

* gem json or yajl-ruby, or multi_json (if `JsonResponse` or
  `JsonRequest` middlewares are used)

## INSTALLATION:

``` shell
gem install rest-more
```

Or if you want development version, put this in Gemfile:

``` ruby
gem 'rest-more', :git => 'git://github.com/godfat/rest-more.git',
                 :submodules => true
```

## SYNOPSIS:

### Dropbox example:

Check out their
[API documentation](https://www.dropbox.com/developers/reference/api)
for a complete reference, and [RC::Dropbox][] for built-in APIs.

``` ruby
require 'rest-more'

d = RC::Dropbox.new :root => 'sandbox',
                    :consumer_key => 'key',
                    :consumer_secret => 'secret',
                    :log_method => method(:puts)

# Redirect the user to:
d.authorize_url!

# After the user authorized, then we can do this to obtain the access token:
d.authorize!

# Then we could call the API:
p [d.me, d.ls]
```

### Facebook example:

Check out their
[Graph API documentation](https://developers.facebook.com/docs/graph-api/reference/v2.0/)
for a complete reference, and [RC::Facebook][] for built-in APIs.
[RC::Facebook::RailsUtil][] for Facebook is also available.

[RC::Facebook::RailsUtil]: lib/rest-core/client/facebook/rails_util.rb

``` ruby
require 'rest-more'

f = RC::Facebook.new :app_id => '123',
                     :secret => 'secret',
                     :access_token => 'if you have the token',
                     :log_method => method(:puts)

redirect_uri = 'http://example.com'
scope = 'public_profile,email'

# Redirect the user to:
f.authorize_url(:redirect_uri => redirect_uri, :scope => scope)

# After the user authorized, then we can do this to obtain the access token:
f.authorize!(:redirect_uri => redirect_uri, :code => 'code')

# Then we could call the API:
p [f.me, f.get('me/posts')]
```

### Github example:

Check out their
[API documentation](https://developer.github.com/v3/)
for a complete reference, and [RC::Github][] for built-in APIs.

``` ruby
require 'rest-more'

g = RC::Github.new :access_token => 'if you have the token',
                   :log_method => method(:puts)

p [g.me, g.get('users/godfat')]
p g.all('users/godfat/repos').size # get all repositories across all pages

```

### Instagram example:

Check out their
[Developer documentation](http://instagram.com/developer/)
for a complete reference, and [RC::Instagram][] for built-in APIs.

``` ruby
require 'rest-more'

i = RC::Instagram.new :client_id => 'id',
                      :client_secret => 'secret',
                      :log_method => method(:puts)

redirect_uri = 'http://example.com'

# Redirect the user to:
i.authorize_url(:redirect_uri => redirect_uri)

# After the user authorized, then we can do this to obtain the access token:
i.authorize!(:redirect_uri => redirect_uri, :code => 'code')

# Then we could call the API:
p i.me
```

### Linkedin example:

Check out their
[API documentation](http://developer.linkedin.com/documents/linkedin-api-resource-map)
for a complete reference, and [RC::Linkedin][] for built-in APIs.

``` ruby
require 'rest-more'

l = RC::Linkedin.new :consumer_key => 'key',
                     :consumer_secret => 'secret',
                     :log_method => method(:puts)

# Redirect the user to:
l.authorize_url!

# After the user authorized, then we can do this to obtain the access token:
l.authorize!(:oauth_verifier => 'oauth_verifier')

# Then we could call the API:
p l.me
```

### Twitter example:

Check out their
[REST API documentation](https://dev.twitter.com/docs/api/1.1)
for a complete reference, and [RC::Twitter][] for built-in APIs.

``` ruby
require 'rest-more'

t = RC::Twitter.new :consumer_key => 'key',
                    :consumer_secret => 'secret',
                    :log_method => method(:puts)

# Redirect the user to:
t.authorize_url!

# After the user authorized, then we can do this to obtain the access token:
t.authorize!(:oauth_token => 'oauth_token',
             :oauth_verifier => 'oauth_verifier')

# Then we could call the API:
p [t.me, t.statuses('godfat')]
p t.tweet('Aloha!', File.open('screen.png')) # Image is optional
```

### Example codes:

Runnable example is at: [example/simple.rb][]. Please see [slides][] from
[rubyconf.tw/2011][rubyconf.tw] for concepts.

[example/simple.rb]: https://github.com/godfat/rest-more/blob/master/example/simple.rb
[slides]: http://www.godfat.org/slide/2011-08-27-rest-core.html
[rubyconf.tw]: http://rubyconf.tw/2011/#6

## Concurrent HTTP Requests:

Inherited from [rest-core][], you can do concurrent requests quite easily.
Here's a very quick example of getting Facebook users' names for UID 4 and 5:

``` ruby
require 'rest-more'
facebook = RC::Facebook.new(:log_method => method(:puts))
puts "httpclient with threads doing concurrent requests"
a = [facebook.get('4'), facebook.get('5')]
puts "It's not blocking... but doing concurrent requests underneath"
p a.map{ |r| r['name'] } # here we want the values, so it blocks here
puts "DONE"
```

If you prefer callback based solution, this would also work:

``` ruby
require 'rest-more'
facebook = RC::Facebook.new(:log_method => method(:puts))
puts "callback also works"
facebook.get('6') do |r|
  p r['name']
end
puts "It's not blocking... but doing concurrent requests underneath"
facebook.wait # we block here to wait for the request done
puts "DONE"
```

Runnable example is at: [example/multi.rb][]. For a detailed demonstration,
see: [Advanced Concurrent HTTP Requests -- Embrace the Future][future]

[example/multi.rb]: https://github.com/godfat/rest-more/blob/master/example/multi.rb
[future]: https://github.com/godfat/rest-core#advanced-concurrent-http-requests----embrace-the-future

## Rails Utilities

To be added. But you can take a look at [Facebook tutorial][] first.

[Facebook tutorial]: doc/facebook.md

## A simple interactive shell with [rib][]:

You need to install [rib][] in order to try this interactive shell:

    gem install rib

Then you can try this by running `rib rest-core`:

    rest-core>> self.site = 'https://api.github.com/users/'
    rest-core>> self.json_response = true
    rest-core>> get 'godfat'

Which is using `RestCore::Universal` for accessing arbitrary websites.

[rib]: https://github.com/godfat/rib

## rest-more users:

* [s2sync](https://github.com/brucehsu/s2sync)
* [s2sync_web](https://github.com/brucehsu/s2sync_web)

## Powered sites:

* [Codementor](https://www.codementor.io/)
* [PicCollage](http://pic-collage.com/)

## CHANGES:

* [CHANGES](CHANGES.md)

## CONTRIBUTORS:

* ayaya (@ayamomiji)
* Fumin (@fumin)
* khoa nguyen (@khoan)
* Lin Jen-Shin (@godfat)
* Yun-Yan Chi (@jaiyalas)

## LICENSE:

Apache License 2.0

Copyright (c) 2011-2015, Lin Jen-Shin (godfat)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   <http://www.apache.org/licenses/LICENSE-2.0>

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGES.md
# CHANGES

## rest-more 3.4.0 -- 2016-02-04

* `RC::Github` would now follow redirect for 5 times by default.
* `RestCore::Github::Error::ClientError` was never used and removed.
* `RestCore::Github::Error::ServerError` was also removed and favouring
  just `RestCore::Github::Error`.
* Adopt rest-core 4.0.0

## rest-more 3.3.4 -- 2015-10-15

* [`Dropbox`] Updated content API server. Thanks khoa nguyen (@khoan)
* [`Github`] Introduced `RC::Github::Error` replacing `RuntimeError`

## rest-more 3.3.3 -- 2015-02-26

* Try to adopt StackExchange's inconsistency

## rest-more 3.3.2 -- 2015-01-05

* Added RC::StackExchange#authorized?
* Removed RC::Firebase from autoloading.

## rest-more 3.3.1 -- 2014-10-08

* Made `RC::Facebook#authorize_url` accepts the second argument as the option.
  #5 khoa nguyen (@khoan)

* The same fix applied to `RC::Facebook#authorize!`. #5 khoa nguyen (@khoan)

* `RC::Github` now accepts `:client_id`, `:client_secret` as options, and
  `RC::Github#authorize_url` and `RC::Github#authorize!` and
  `RC::Github#authorized?` are now implemented. Lin Jen-Shin (@godfat)

## rest-more 3.3.0 -- 2014-08-25

* Adopted rest-core 3.3.0

## rest-more 3.2.0 -- 2014-06-27

* Removed RC::Firebase because it's extracted into [rest-firebase][]
* A lot of methods now accept callback properly.

[rest-firebase]: https://github.com/CodementorIO/rest-firebase

### Changes for RC::Github

* Added `RC::Github#all` to grab resources through all pages automatically
  and concurrently. e.g. `RC::Github.new.all('users/godfat/repos')`

### Changes for RC::Linkedin

* Added `RC::Linkedin#profile` to access someone's profile easier.

### Added RC::StackExchange

## rest-more 3.1.0 -- 2014-05-09

### Changes for RC::Firebase

* Now it would properly follow redirect.
* Now it would raise an error if we're not giving a secret to RC::Firebase.
* Now we have RC::Firebase::Error and its subclasses.
* Introduced RC::Firebase::EventSource which would parse the data instead of
  simply giving the string for onmessage callback.

## rest-more 3.0.0 -- 2014-05-04

* Adopted latest rest-core
* Added `RC::Instagram`, thanks @khoan
* Added `RC::Firebase`

## rest-more 2.0.4 -- 2013-04-13

* [Facebook::RailsUtil] Since now that we cannot retrieve the access token
  from cookie directly, we need to exchange it with the code in cookie,
  and then save it to our own storage. This fixed some issues with Facebook
  JavaScript SDK used together.

* [Facebook::RailsUtil] Changed the order checking access token, so that
  we don't have to exchange the access token every time.

* [Twitter] Adopted Twitter API 1.1, thanks @khoan.

## rest-more 2.0.3 -- 2013-04-01

* Fixed compatibility with Ruby 2.0.
* Adopt rest-core 2.0.3.

## rest-more 2.0.2 -- 2013-02-07

* Updated Facebook OAuth URL from https://graph.facebook.com/oauth/authorize
  to https://www.facebook.com/dialog/oauth

## rest-more 2.0.0 -- 2012-10-31

* Adopted to rest-core 2.0.0, a whole new generation.
* We no longer test against Ruby 1.9.2- and Rails 2-

### Incompatible changes

* Bing (no longer useable) and Mixi (anyone?) clients are removed.
* Option for `:json_decode` is renamed to `:json_response` in rest-core.

### Enhancement

* [Facebook::RailsUtil] Now we cache `@rc_facebook_normalized_request_uri`
  to avoid paring broken URI from Facebook. Facebook is returning some
  URI which are not properly escaped.

## rest-more 1.0.2 -- 2012-07-13

* [Facebook::RailsUtil] Change the redirect log level from WARN to INFO.
* [Facebook::RailsUtil] Since Facebook would return correct URL now,
  we don't have to try to use `URI.encode` anymore. Actually, that
  causes bugs.

* [Twitter::RailsUtil] Change the redirect log level from WARN to INFO.

* [Flurry] Has been removed because it's too hard to maintain and
  make it right.

## rest-more 1.0.1 -- 2012-05-14

* [RailsUtilUtil] Fix for DalliStore

## rest-more 1.0.0 -- 2012-03-17

### Enhancement

All clients are now asynchrony-aware. Asynchrony support is added in
rest-core >=1.0.0. Whenever you pass a callback block, the response
would then be passed to the block. Error handling is different, too.
In synchronous style, usually an exception would be raised, but in
asynchronous style, the exception would be passed to the block instead
of being raised.

* [`Dropbox`] Added Dropbox support.
* [`Bing::Error`] Added `code` method to get the original error code.
* [`Twitter::Error`] Added `code` method to get HTTP status code.
* [`Facebook::Error`] Instead of passing `[true]`, pass error messages.
* [`Facebook`] Pages related API is reimplemented. Passing a block would
  result a serious call to the callback block, ending with a nil.
* [`Config`] Make sure the default attributes module is there even if
  the config file isn't presented.

## rest-more 0.8.0 -- 2011-11-29

### Incompatible changes

* [`Facebook::RailsUtil`] Some methods are renamed. For example,
  `rc_facebook_write_rg_fbs` is renamed to `rc_facebook_write_fbs`.
  All renamed methods are considered private, so we're not listing them here.

* [`Facebook::RailsUtil`] `rc_facebook_storage_key` is changed to:
  `"rc_facebook_#{rc_facebook.app_id}"`, your users might need to
  login again in order to save the access_token into the new place.

### Bugs fixes

* [`Github`] Fixed usage of `Oauth2Query`.

* [`Facebook`] Now we're using POST in `authorize!` to exchange the
  access_token with the code instead of GET. If we're using GET,
  we would run into a risk where a user might use the code to
  get other people's access_token via the cache. Using POST would
  prevent this because POSTs are not cached.

* [`Facebook::RailsUtil`] Fixed a serious bug. The bug would jump up if
  you're using :write_session or :write_cookies or :write_handler along
  with :auto_authorize, for example:
  `rc_facebook_setup(:auto_authorize => true, :write_session => true)`
  The problem is that `Facebook::RailsUtil` is not removing the invalid

lib/rest-core/client/flurry.rb

require 'rest-core'

require 'time' # for Time.parse

# http://wiki.flurry.com
RestCore::Flurry = RestCore::Builder.client(:apiKey, :apiAccessCode) do
  s = self.class # this is only for ruby 1.8!
  use s::Timeout       , 10

  use s::DefaultSite   , 'http://api.flurry.com/'
  use s::DefaultHeaders, {'Accept' => 'application/json'}
  use s::DefaultQuery  , {}

  use s::CommonLogger  , nil
  use s::Cache         , nil, 600 do
    use s::ErrorHandler, lambda{ |env|
      if env[s::ASYNC]
        if env[s::RESPONSE_BODY].kind_of?(::Exception)
          env
        else
          env.merge(s::RESPONSE_BODY =>
                      RuntimeError.new(env[s::RESPONSE_BODY]['message']))
        end
      else
        raise env[s::RESPONSE_BODY]['message']
      end}
    use s::ErrorDetectorHttp
    use s::JsonDecode  , true
  end
end

module RestCore::Flurry::Client
  # see: http://wiki.flurry.com/index.php?title=AppInfo
  # >> f.app_info
  # => {"@platform"=>"iPhone", "@name"=>"PicCollage",
  #     "@createdDate"=>"2011-07-24", "@category"=>"Photography",
  #     "@version"=>"1.0", "@generatedDate"=>"9/15/11 7:08 AM",
  #     "version"=>[{"@name"=>"2.1", ...
  def app_info query={}, opts={}
    get('appInfo/getApplication', query, opts)
  end

  # see: http://wiki.flurry.com/index.php?title=EventMetrics
  # >> f.event_names
  # => ["Products", "Save To Photo Library", ...]
  def event_names query={}, opts={}
    event_summary(query, opts).keys
  end

  # see: http://wiki.flurry.com/index.php?title=EventMetrics
  # >> f.event_summary({}, :days => 7)
  # => {"Products" => {"@usersLastWeek"  => "948" ,
  #                    "@usersLastMonth" => "2046",
  #                    "@usersLastDay"   => "4"   , ...}}
  def event_summary query={}, opts={}
    array2hash(get('eventMetrics/Summary',
                   *calculate_query_and_opts(query, opts))['event'],
               '@eventName')
  end

  # see: http://wiki.flurry.com/index.php?title=EventMetrics
  # >> f.event_metrics('Products', {}, :days => 7)
  # => [["2011-11-23", {"@uniqueUsers"   => "12"     ,
  #                     "@totalSessions" => "108392" ,
  #                     "@totalCount"    => "30"     ,
  #                     "@duration"      => "9754723"}],
  #     ["2011-11-22", {...}]]
  def event_metrics name, query={}, opts={}
    array2hash(get('eventMetrics/Event',
                   *calculate_query_and_opts(
                      {'eventName' => name}.merge(query), opts))['day'],
               '@date').sort{ |a, b| b <=> a }
  end

  # see: http://wiki.flurry.com/index.php?title=AppMetrics
  # >> f.metrics('ActiveUsers', {}, :weeks => 4)
  # => [["2011-09-19",  6516], ["2011-09-18", 43920], ["2011-09-17", 45412],
  #     ["2011-09-16", 40972], ["2011-09-15", 37587], ["2011-09-14", 34918],
  #     ...]
  def metrics path, query={}, opts={}
    get("appMetrics/#{path}", *calculate_query_and_opts(query, opts)
      )['day'].map{ |i| [i['@date'], i['@value'].to_i] }.reverse
  end

  # >> f.weekly(f.metrics('ActiveUsers', {}, :weeks => 4))
  # => [244548, 270227, 248513, 257149]
  def weekly array
    start = Time.parse(array.first.first, nil).to_i
    array.group_by{ |(date, value)|
      current = Time.parse(date, nil).to_i
      - (current - start) / (86400*7)
    # calling .last to discard week numbers created by group_by
    }.sort.map(&:last).map{ |week|
      week.map{ |(_, num)| num }.inject(&:+) }
  end

  # >> f.sum(f.weekly(f.metrics('ActiveUsers', {}, :weeks => 4)))
  # => [1020437, 775889, 505662, 257149]
  def sum array
    reverse = array.reverse
    (0...reverse.size).map{ |index|
      reverse[1, index].inject(reverse.first, &:+)
    }.reverse
  end

  def default_query
    {'apiKey'        => apiKey       ,
     'apiAccessCode' => apiAccessCode}
  end

  private
  def calculate_query_and_opts query, opts
    days = opts[:days] || (opts[:weeks]  && opts[:weeks] * 7)   ||
                          (opts[:months] && opts[:months] * 30) || 7

    startDate = query[:startDate] || (Time.now + 86400 - 86400*days).
      strftime('%Y-%m-%d')

    endDate   = query[:endDate]   || Time.now.
      strftime('%Y-%m-%d')

    [query.merge(:startDate => startDate,
                 :endDate   => endDate),
     opts.reject{ |k, _| [:days, :weeks, :months].include?(k) }]
  end

  def array2hash array, key
    array.inject({}){ |r, i|
      r[i[key]] = i.reject{ |k, _| k == key }
      r }
  end
end

RestCore::Flurry.send(:include, RestCore::Flurry::Client)
require 'rest-core/client/flurry/rails_util' if
  Object.const_defined?(:Rails)
